### PR TITLE
feature/se-9-upload-CMDI

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|-----:|-----:|-------:|
| ETA                  |  26.4  |    0 |     0 |      0 |        0 |
| Developing      |  8.6+8+8+5.4+1+1  |    0 |    0 |      0 |         0 |
| Review             |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -   |   -     |        0 |
## Problem description
For the admin must be a option to upload CMDI file in the Submission UI. The CMDI file must be stored in the:
1. METADATA bundle - the admin has selected the option to upload the CMDI file
2. ORIGINAL bundle - the hasn't selected the option to upload the CMDI file

Option: 
image

1. Option is selected and the CMDI file is uploaded ->  the file won't be visible in the Item View
2. Option is not selected and the CMDI file is uploaded -> the file will be visible in the Item View

- [x] Define a new local metadata type hasCMDI
- [x] Create a test for a new local metadata
- [x] Define a new input field (checkbox) for the admin - is uploading a cmdi file.
- [x] Every file will be pushed to the bundle ORIGINAL then the CMDI file will be moved to the METADATA bundle by the MetadataValidation class.
- [x] If the user is admin and has selected option to upload CMDI file, the CMDI file must be stored in the bundle METADATA. If the user hasn't selected option to upload CMDI file but is uploading the CMDI file, the CMDI file must be stored in the bundle ORIGINAL. This check will be done by the MetadataValidation in the method of the new class _CMDIFileBundleMaintainer.updateCMDIFileBundle_.
- [x] Create a test for uploading CMDI file.
- [x] Create a tests for maintaining CMDI file in the correct bundles.

### Two requests problem
After FIRST save/deposit/upload operation in the Submission UI are called two requests to the server:
1. PATCH to create a new Item to the DB
3. POST to update filled metadata to the ITEM or upload a file
Then is called only POST request. 

If the first operation was uploading the file with filled up item's metadata, the UploadValidation doesn't see filled metadata values because the PATCH and POST requests are processing parallel. The POST request cannot see the item's metadata because PATCH is not to fast to save the item to the database before the POST. Then the POST request cannot see if the the admin has selected the option to upload CMDI file.
**Solution**:
When the user uploads a file all two request are called and the file (CMDI or something else) is stored to the bundle ORIGINAL. After a Deposit/Save/Save for later is called the POST request and is called to check if the CMDI file is in the right bundle - if not - move it to the right bundle.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java [hunk 1]
```diff
@@ -539,11 +539,11 @@ public void move(Context context, Item item, Collection from, Collection to, boo
         throws SQLException, AuthorizeException, IOException;
 
     /**
-     * Check the bundle ORIGINAL to see if there are any uploaded files
+     * Check the bundle ORIGINAL and METADATA to see if there are any uploaded files
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/content/service/ItemService.java:542]: @milanmajchrak this is dpace-core? Anything to do with CMDI will never get to upstream

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 2]
```diff
@@ -96,6 +98,14 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
                             "/" + WorkspaceItemRestRepository.OPERATION_PATH_SECTIONS + "/" + config.getId() + "/" +
                                 input.getFieldName());
                     }
+                    if ("local.hasCMDI".equals(fieldName)) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:101]: Could we use a variable? Possibly configurable

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 3]
```diff
@@ -96,6 +98,14 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
...
+                        try {
+                            Context context = ContextUtil.obtainCurrentRequestContext();
+                            CMDIFileBundleMaintainer.updateCMDIFileBundle(context, obj.getItem(), mdv);
+                        } catch (Exception e) {
+                            e.printStackTrace();
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:106]: Is this really used also elsewhere?

File: dspace/config/registries/local-types.xml [hunk 4]
```diff
@@ -30,4 +30,10 @@
     <scope_note>Date issued used where the point in time is uncertain; eg. a range.</scope_note>
   </dc-type>
 
+  <dc-type>
+    <schema>local</schema>
+    <element>hasCMDI</element>
+    <scope_note>Indicates uploaded cmdi file</scope_note>
```
Comment (vidiecan) [inline dspace/config/registries/local-types.xml:36]: maybe add also the values that could be there, true/false?

**Comment by vidiecan** [comment]
@milanmajchrak very thorough description, please add a screenshot how it looks/should look

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:106]: No, but for testing purposes I separated it.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java (hunk 2 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java]: Added constant: https://github.com/dataquest-dev/DSpace/blob/feature/se-9-upload-CMDI/dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java#L47

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java (hunk 1 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/content/service/ItemService.java]: Changed

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java [hunk 5]
```diff
@@ -539,14 +539,15 @@ public void move(Context context, Item item, Collection from, Collection to, boo
         throws SQLException, AuthorizeException, IOException;
 
     /**
-     * Check the bundle ORIGINAL to see if there are any uploaded files
+     * Check the bundle ORIGINAL or METADATA to see if there are any uploaded files
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/content/service/ItemService.java:542]: are you sure, cannot bundleName be anything?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        List<Bundle> bundleMETADATA = itemService.getBundles(item, Constants.METADATA_BUNDLE_NAME);
+        List<Bundle> bundleORIGINAL = itemService.getBundles(item, Constants.CONTENT_BUNDLE_NAME);
+
+        String targetBundle = "";
+        List<Bundle> bundleToProcess = null;
+
+        if (!bundleMETADATA.isEmpty() && mdv.isEmpty()) {
+            targetBundle = Constants.CONTENT_BUNDLE_NAME;
+            bundleToProcess = bundleMETADATA;
+        } else if (!bundleORIGINAL.isEmpty() && !mdv.isEmpty() && "yes".equals(mdv.get(0).getValue())) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java:59]: "yes"?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 7]
```diff
@@ -96,6 +100,14 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
                             "/" + WorkspaceItemRestRepository.OPERATION_PATH_SECTIONS + "/" + config.getId() + "/" +
                                 input.getFieldName());
                     }
+                    if (LOCAL_METADATA_HAS_CMDI.equals(fieldName)) {
+                        try {
+                            Context context = ContextUtil.obtainCurrentRequestContext();
+                            CMDIFileBundleMaintainer.updateCMDIFileBundle(context, obj.getItem(), mdv);
+                        } catch (Exception e) {
+                            e.printStackTrace();
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:108]: ?

File: dspace/config/registries/local-types.xml (hunk 4 above)
Comment (vidiecan) [reply dspace/config/registries/local-types.xml:36]: ?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java (hunk 6 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java:59]: added constant

File: dspace-api/src/main/java/org/dspace/content/service/ItemService.java (hunk 5 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/content/service/ItemService.java:542]: changed comment

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java (hunk 7 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:112]: added log.error

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        } else if (!bundleORIGINAL.isEmpty() && !mdv.isEmpty() &&
+                StringUtils.equals(hasCMDYesOptionValue, mdv.get(0).getValue())) {
+            targetBundle = Constants.METADATA_BUNDLE_NAME;
+            bundleToProcess = bundleORIGINAL;
+        }
+
+        for (Bundle bundle : CollectionUtils.emptyIfNull(bundleToProcess)) {
+            for (Bitstream bitstream : bundle.getBitstreams()) {
+                // change bundle only for cmdi file
+                if (bitstream.getName().toLowerCase().endsWith(".cmdi")) {
```
Comment (mrovnanik) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java:70]: exit loop if not cmdi

**Review by mrovnanik — COMMENTED** [review]
please, implement the small change in comment

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java (hunk 8 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/CMDIFileBundleMaintainer.java]: done

**Review by mrovnanik — APPROVED** [review]
thanks

**Pull request merged**